File: src/gestorAplicacion/servicios/Servicio.java
/* Autores: Diego Andres Gracia Granados, Daniel Giraldo Vanegas, Elian David Velandia Riveros, Juan Camilo Gutierrez Martinez
y Santiago Arboleda Acevedo */
package gestorAplicacion.servicios;

import gestorAplicacion.administracion.HistoriaClinica;
import gestorAplicacion.personas.Paciente;

import java.io.Serializable;
import java.util.ArrayList;

//Clase Abstracta destinada a herencia de servicios del hospital
public abstract class Servicio implements Serializable {

    //Atributos
    protected static long generadorID = 1000;
    protected final long idServicio;
    protected Paciente paciente;
    private boolean estadoPago;

    //Constructor
    public Servicio(Paciente paciente) {
        this.idServicio = generadorID++;
        this.paciente = paciente;
        this.estadoPago = false;
    }

    //Métodos

    /* Método abstracto que debe implementarse en las clases hijas para
    validar el pago
     */
    public abstract void validarPago(Paciente paciente, long idServicio);

    /* Metodo abstracto para imprimir la descripcion del servicio.
    Se usa este metodo porque el toString se utiliza en otra funcionalidad.
     */
    public abstract String descripcionServicio();

    //Método estático para obtener servicios sin pagar
    public static ArrayList<Servicio> obtenerServiciosSinPagar(Paciente paciente) {
        HistoriaClinica historiaClinicaPaciente = paciente.getHistoriaClinica();
        ArrayList<Servicio> serviciosSinPagar = new ArrayList<>();

        // Obtiene todos los servicios brindados al paciente
        serviciosSinPagar.addAll(historiaClinicaPaciente.getHistorialCitas());
        serviciosSinPagar.addAll(historiaClinicaPaciente.getListaFormulas());
        if (paciente.getHabitacionAsignada() != null)
            serviciosSinPagar.add(paciente.getHabitacionAsignada());
        serviciosSinPagar.addAll(historiaClinicaPaciente.getHistorialVacunas());

        // Filtra los servicios pagados
        serviciosSinPagar.removeIf(Servicio::isEstadoPago);

        return serviciosSinPagar;
    }

    //Getters y Setters
    public boolean isEstadoPago() {
        return estadoPago;
    }

    public long getIdServicio() {
        return idServicio;
    }

    public void setEstadoPago(boolean estadoPago) {
        this.estadoPago = estadoPago;
    }

    public Paciente getPaciente() {
        return paciente;
    }

    public void setPaciente(Paciente paciente) {
        this.paciente = paciente;
    }
}

File: Python/src/gestor_aplicacion/servicios/servicio.py
from abc import ABC, abstractmethod


class Servicio(ABC):
    generadorID = 1

    def __init__(self, paciente):
        self._id_servicio = Servicio.generadorID
        Servicio.generadorID += 1
        self._paciente = paciente
        self._estado_pago = False

    @abstractmethod
    def descripcion_servicio(self):
        pass

    @abstractmethod
    def validar_pago(self, paciente, id_servicio):
        pass

    @property
    def id_servicio(self):
        return self._id_servicio

    @property
    def estado_pago(self):
        return self._estado_pago

    @estado_pago.setter
    def estado_pago(self, estado_pago):
        self._estado_pago = estado_pago

    @classmethod
    def obtener_servicios_sin_pagar(cls, paciente):
        historia_clinica_paciente = paciente.historia_clinica
        servicios_sin_pagar = []

        # Obtiene todos los servicios brindados al paciente
        servicios_sin_pagar.extend(historia_clinica_paciente.historial_citas)
        servicios_sin_pagar.extend(historia_clinica_paciente.lista_formulas)
        if paciente.habitacion_asignada is not None:
            servicios_sin_pagar.append(paciente.habitacion_asignada)
        servicios_sin_pagar.extend(historia_clinica_paciente.historial_vacunas)

        # Filtra los servicios pagados
        servicios_sin_pagar = [servicio for servicio in servicios_sin_pagar if not servicio.estado_pago]
        servicios_sin_pagar.sort(key=lambda servicio: servicio.id_servicio)

        return servicios_sin_pagar
